src/builtins.rs

```diff
@@ -0,0 +1,686 @@
+                        map.insert(name.to_string(), Gc::new(Tree::from_concrete(
+                            NixValue::Bool(value)
+                        )));
+                    }
+                    Ok(Gc::new(NixValue::Map(map)))
+                }
+                NODE_IDENT => {
+                    Ok(Gc::new(NixValue::Map(HashMap::new())))
+                }
+                x => Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+            }
+        },
+        x => Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+    }
+    "length" ; Length => |param: Gc<Tree>| match param.eval()?.borrow() {
+        NixValue::List(x) => Ok(Gc::new(NixValue::Integer(x.len() as i64))),
+        x => Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+    }
+    "head" ; Head => |param: Gc<Tree>| param.eval()?.as_list()?[0].eval()
+    "tail" ; Tail => |param: Gc<Tree>| Ok(Gc::new(
+        NixValue::List(param.eval()?.as_list()?[1..].to_vec())))
+    "stringLength" ; StringLength => |param: Gc<Tree>| Ok(Gc::new(
+        NixValue::Integer(param.eval()?.as_str()?.len() as i64)))
+    "getEnv" ; GetEnv => |param: Gc<Tree>| Ok(Gc::new(NixValue::Str(
+        std::env::var(param.eval()?.as_str()?).unwrap_or_else(|_| "".into()))))
+    "pathExists" ; PathExists => |param: Gc<Tree>| match param.eval()?.borrow() {
+        NixValue::Path(x, y) => Ok(Gc::new(NixValue::Bool(expand_path(x.clone(), y.clone()).ok_or(EvalError::Parsing)?.exists()))),
+        NixValue::Str(x) => Ok(Gc::new(NixValue::Bool(PathBuf::from_str(x).map_err(|_| EvalError::Parsing)?.exists()))),
+        x => Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+    }
+    "unsafeDiscardStringContext" ; UnsafeDiscardStringContext => |param: Gc<Tree>| {
+        param.eval()
+    }
+    // FIXME: missing functionality
+    "addErrorContext" ; AddErrorContext => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::AddErrorContext1(param)))))
+    }
+    "addErrorContext <?>" ; AddErrorContext1(_0: GcLazy) => |param: Gc<Tree>, _x| {
+        param.eval()
+    }
+    // FIXME: missing functionality
+    "trace" ; Trace => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::Trace1(param)))))
+    }
+    "trace <?>" ; Trace1(_0: GcLazy) => |param: Gc<Tree>, _x| {
+        param.eval()
+    }
+    // FIXME: missing functionality
+    "tryEval" ; TryEval => |param: Gc<Tree>| {
+        let mut map = HashMap::new();
+        let tmp = param.eval();
+        map.insert("success".to_string(), Gc::new(Tree::from_concrete(
+            NixValue::Bool(tmp.is_ok())
+        )));
+        match tmp {
+            Ok(x) => map.insert("value".to_string(), Gc::new(
+                Tree::from_concrete({
+                    let a: &NixValue = x.borrow();
+                    a.clone()
+                }))),
+            Err(_) => map.insert("value".to_string(), Gc::new(
```

Comment:
Yep. We'll want a error type for user-generated errors. This will also help if we later to do the red squiggley diagnostic thing to show exactly which user input (e.g. a nixos option value) is causing an assertion to fail somewhere else.